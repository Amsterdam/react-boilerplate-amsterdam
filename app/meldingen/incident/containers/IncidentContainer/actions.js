/*
 *
 * IncidentContainer actions
 *
 */

import {
  SET_INCIDENT,
  CREATE_INCIDENT,
  CREATE_INCIDENT_SUCCESS,
  CREATE_INCIDENT_ERROR
} from './constants';

export function setIncident(incident) {
  return {
    type: SET_INCIDENT,
    incident
  };
}

export function createIncident(incident) {
  return {
    type: CREATE_INCIDENT,
    incident
  };
}

export function createIncidentSuccess(incident) {
  return {
    type: CREATE_INCIDENT_SUCCESS,
    incident
  };
}

export function setIncidentError(message) {
  return {
    type: CREATE_INCIDENT_ERROR,
    message
  };
}
