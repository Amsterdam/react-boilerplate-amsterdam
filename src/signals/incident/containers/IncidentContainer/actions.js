/*
 *
 * IncidentContainer actions
 *
 */

import {
  SET_INCIDENT,

  CREATE_INCIDENT,
  CREATE_INCIDENT_SUCCESS,
  CREATE_INCIDENT_ERROR,

  GET_CLASSIFICATION,
  GET_CLASSIFICATION_SUCCESS,
  GET_CLASSIFICATION_ERROR
} from './constants';

export function setIncident(incident) {
  return {
    type: SET_INCIDENT,
    payload: { ...incident }
  };
}

export function createIncident(incident, wizard) {
  return {
    type: CREATE_INCIDENT,
    payload: {
      incident: { ...incident },
      wizard: { ...wizard }
    }
  };
}

export function createIncidentSuccess(incident) {
  return {
    type: CREATE_INCIDENT_SUCCESS,
    payload: { ...incident }
  };
}

export function createIncidentError() {
  return {
    type: CREATE_INCIDENT_ERROR
  };
}

export function getClassification(text) {
  return {
    type: GET_CLASSIFICATION,
    payload: { text }
  };
}

export function getClassificationSuccess(classification) {
  return {
    type: GET_CLASSIFICATION_SUCCESS,
    payload: { ...classification }
  };
}

export function getClassificationError(classification) {
  return {
    type: GET_CLASSIFICATION_ERROR,
    payload: { ...classification }
  };
}
