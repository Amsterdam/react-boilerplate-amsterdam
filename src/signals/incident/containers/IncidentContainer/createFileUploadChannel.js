import {
  buffers,
  eventChannel,
  END
} from 'redux-saga';

function createFileUploadChannel(endpoint, file, id) {
  return eventChannel((emitter) => {
    const formData = new window.FormData();
    formData.append('signal_id', id);
    formData.append('image', file);

    const xhr = new window.XMLHttpRequest();

    xhr.upload.onprogress = (e) => {
      console.log('progress', e);
      const progress = e.loaded / e.total;
      emitter({ progress });
    };

    xhr.onload = () => {
      // upload success
      if (xhr.readyState === 4 && xhr.status === 202 && xhr.statusText === 'Accepted') {
        console.log('success!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!');
        emitter({ success: true });
        emitter(END);
      } else {
        emitter({ err: new Error('Upload failed') });
        emitter(END);
      }
    };

    xhr.open('POST', endpoint, true);
    xhr.send(formData);

    return () => {
      // xhr.upload.removeEventListener('progress', onProgress);
      // xhr.upload.removeEventListener('error', onFailure);
      // xhr.upload.removeEventListener('abort', onFailure);
      // xhr.onreadystatechange = null;
      // xhr.abort();
    };
  }, buffers.sliding(2));
}

export default createFileUploadChannel;
