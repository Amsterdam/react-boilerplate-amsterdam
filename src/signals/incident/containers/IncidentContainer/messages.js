/*
 * IncidentContainer Messages
 *
 * This contains all the text for the IncidentContainer component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  header: {
    id: 'app.containers.IncidentContainer.header',
    defaultMessage: 'This is IncidentContainer container !',
  },
});
