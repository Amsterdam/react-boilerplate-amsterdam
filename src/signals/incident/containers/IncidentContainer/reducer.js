/*
 *
 * IncidentContainer reducer
 *
 */

import { fromJS } from 'immutable';
import {
  SET_INCIDENT,

  CREATE_INCIDENT,
  CREATE_INCIDENT_SUCCESS,
  CREATE_INCIDENT_ERROR,

  GET_CLASSIFICATION_SUCCESS,
  GET_CLASSIFICATION_ERROR,

  SET_PRIORITY,
  SET_PRIORITY_SUCCESS
} from './constants';

export const initialState = fromJS({
  incident: {
    location: {
      address: {
        openbare_ruimte: 'Diemerparklaan',
        huisnummer: '16',
        huisletter: '',
        huisnummer_toevoeging: '',
        postcode: '1087GE',
        woonplaats: 'Amsterdam'
      },
      buurt_code: 'M35c',
      stadsdeel: 'M',
      geometrie: {
        type: 'Point',
        coordinates: [
          4.995174407958985,
          52.35451699670258
        ]
      }
    },
    subcategory: 'Overlast op het water - snel varen',
    category: 'Overlast op het water',
    description: 'snel boot',
    phone: '020654321',
    email: 'a@b.com',
    extra_boten_snelheid_rondvaartboot: 'Ja',
    extra_boten_snelheid_rederij: 'Aemstelland',
    datetime: 'Nu',

    incident_date: 'Vandaag',
    incident_time_hours: 9,
    incident_time_minutes: 0,
    priority: 'normal'
  },
  priority: {}
});

function incidentContainerReducer(state = initialState, action) {
  switch (action.type) {
    case SET_INCIDENT:
      return state
        .set('incident', fromJS({
          ...state.get('incident').toJS(),
          ...action.payload
        }));

    case CREATE_INCIDENT:
      return state
        .set('loading', true)
        .set('error', false)
        .set('incident', fromJS({
          ...state.get('incident').toJS(),
          id: null
        }));

    case CREATE_INCIDENT_SUCCESS:
      return state
        .set('loading', false)
        .set('incident', fromJS({
          ...(initialState.get('incident').toJS()),
          id: action.payload.id,
          category: action.payload.category.main,
          subcategory: action.payload.category.sub
        }));

    case CREATE_INCIDENT_ERROR:
      return state
        .set('error', true)
        .set('loading', false);

    case GET_CLASSIFICATION_SUCCESS:
    case GET_CLASSIFICATION_ERROR:
      return state
        .set('incident', fromJS({
          ...state.get('incident').toJS(),
          ...action.payload
        }));

    case SET_PRIORITY:
      return state
        .set('priority', fromJS({
          ...action.payload
        }));

    case SET_PRIORITY_SUCCESS:
      return state
        .set('priority', fromJS({}));

    default:
      return state;
  }
}

export default incidentContainerReducer;
