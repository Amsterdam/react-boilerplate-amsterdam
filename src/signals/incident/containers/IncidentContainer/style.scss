.incident-container {
  &:before {
    content: "yooooooo";
    color: white;
    background-color: #ec0000;
    line-height: 22px;
    display: block;
    padding: 10px;
    margin-top: -20px;
  }
}
