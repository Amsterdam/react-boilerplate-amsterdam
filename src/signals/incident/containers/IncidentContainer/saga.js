import { all, call, put, take, /* select, */ takeLatest, takeEvery } from 'redux-saga/effects';
import { replace } from 'react-router-redux';
import request from 'utils/request';

import CONFIGURATION from 'shared/services/configuration/configuration';
import { CREATE_INCIDENT, GET_CLASSIFICATION, UPLOAD_REQUEST } from './constants';
import {
  createIncidentSuccess,
  createIncidentError,
  getClassificationSuccess,
  getClassificationError,
  uploadRequest,
  uploadProgress,
  uploadSuccess,
  uploadFailure
} from './actions';
import { showGlobalError } from '../../../../containers/App/actions';

import mapControlsToParams from '../../services/map-controls-to-params';
import fileUploadChannel from '../../services/file-upload-channel';
import setClassification from '../../services/set-classification';

// import makeSelectIncidentContainer from './selectors';

export function* getClassification(action) {
  const requestURL = `${CONFIGURATION.API_ROOT}signals_mltool/predict`;

  try {
    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({
        text: action.payload.text
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    });

    yield put(getClassificationSuccess(setClassification(result)));
  } catch (error) {
    yield put(getClassificationError(setClassification()));
  }
}

export function* createIncident(action) {
  const requestURL = `${CONFIGURATION.API_ROOT}signals/signal/`;
  try {
    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(mapControlsToParams(action.payload.incident, action.payload.wizard)),
      headers: {
        'Content-Type': 'application/json'
      }
    });

    if (action.payload.incident.image) {
      yield put(uploadRequest(action.payload.incident.image_file, result.id));
    }
    yield put(createIncidentSuccess(result));
  } catch (error) {
    yield put(createIncidentError());
    // yield put(showGlobalError('CREATE_INCIDENT_FAILED'));
    yield put(replace('/incident/fout'));
  }
}

export function* uploadFile(action) {
  const requestURL = `${CONFIGURATION.API_ROOT}signals/signal/image/`;

  const channel = yield call(fileUploadChannel, requestURL, action.payload.file, action.payload.id);
  const forever = true;
  while (forever) {
    const { progress = 0, error, success } = yield take(channel);
    if (error) {
      yield put(uploadFailure(action.payload.file, error));
      yield put(showGlobalError('UPLOAD_FAILED'));
      return;
    }
    if (success) {
      yield put(uploadSuccess(action.payload.file));
      return;
    }
    yield put(uploadProgress(action.payload.file, progress));
  }
}

function* uploadFileWrapper(action) {
  yield call(uploadFile, action);
}

// Individual exports for testing
export default function* watchIncidentContainerSaga() {
  yield all([
    takeLatest(GET_CLASSIFICATION, getClassification),
    takeLatest(CREATE_INCIDENT, createIncident),
    takeEvery(UPLOAD_REQUEST, uploadFileWrapper)
  ]);
}
