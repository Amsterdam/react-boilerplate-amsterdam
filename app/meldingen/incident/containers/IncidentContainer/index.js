/**
 *
 * IncidentContainer
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
// import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import makeSelectIncidentContainer from './selectors';
import reducer from './reducer';
import saga from './saga';
// import messages from './messages';
import './style.scss';

import IncidentWizard from '../../components/IncidentWizard/Loadable';

export class IncidentContainer extends React.Component { // eslint-disable-line react/prefer-stateless-function
  render() {
    return (
      <div className="incident-container">
        <IncidentWizard />
      </div>
    );
  }
}

IncidentContainer.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  incidentcontainer: makeSelectIncidentContainer(),
});

function mapDispatchToProps(dispatch) {
  return {
    dispatch,
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'incidentContainer', reducer });
const withSaga = injectSaga({ key: 'incidentContainer', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(IncidentContainer);
