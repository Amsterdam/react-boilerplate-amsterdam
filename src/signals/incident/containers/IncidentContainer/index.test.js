import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { fromJS } from 'immutable';

import IncidentContainer from './index';
// import IncidentContainer from './index';

jest.mock('../../components/IncidentWizard', () => () => 'IncidentWizard');

describe('<IncidentContainer />', () => {
  let props;

  beforeEach(() => {
    props = {
      incidentContainer: { incident: {} },
      getClassification: jest.fn(),
      updateIncident: jest.fn(),
      createIncident: jest.fn(),
      isAuthenticated: false
    };
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('rendering', () => {
    // it('should render correctly', () => {
      // const wrapper = shallow(
        // <IncidentContainer {...props} />
      // );
      // expect(wrapper).toMatchSnapshot();
    // });

    it('should render correctly', () => {
      const mockStore = configureStore();
      const store = mockStore(fromJS({ incidentContainer: { incident: {} } }));
      const wrapper = mount(<Provider store={store}><IncidentContainer {...props} /></Provider>);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
