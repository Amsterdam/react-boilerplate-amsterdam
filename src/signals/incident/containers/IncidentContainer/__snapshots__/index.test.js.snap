// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IncidentContainer /> rendering should render correctly 1`] = `
<withReducer(withSaga(Connect(IncidentContainer)))
  createIncident={[MockFunction]}
  getClassification={[MockFunction]}
  incidentContainer={
    Object {
      "incident": Object {},
    }
  }
  isAuthenticated={false}
  updateIncident={[MockFunction]}
/>
`;
